### PR TITLE
perf: update usepolling [DET-4075]

### PR DESCRIPTION
**Pull request by hkang1**
## Description

When the polling function is not an async function there isn’t a reliable way to block the next call. The new API supports async calls so going forward we can reliably block redundant API calls when polling. For the old API, the best way to handle it is to just check to see if the `useRestApi` context has the `isLoading` bit set, which means there is already a request in progress.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/src/pages/TrialDetails.tsx [hunk 1]
```diff
@@ -193,7 +193,9 @@ const TrialDetailsComp: React.FC = () => {
   }, [ hasCheckpointOrValidation, trial?.steps ]);
 
   const pollTrialDetails = useCallback(() => {
+    if (trialResponse.isLoading) return;
```
Comment (hamidzr) [inline webui/react/src/pages/TrialDetails.tsx:196]: blocking: could this be never false? 
could we test (maybe manually) this with say adding a delay to the trialDetail API call? 
from my testing I'm not seeing the expected behavior

Screen Capture_select-area_20200904150040


```
diff --git a/webui/react/src/Dev.ts b/webui/react/src/Dev.ts
index ac3a8ee2..66e7085c 100644
--- a/webui/react/src/Dev.ts
+++ b/webui/react/src/Dev.ts
@@ -7,3 +7,9 @@ if (process.env.IS_DEV) {
   window.dev.api = Api;
   document.title = `[dev] ${document.title}`;
 }
+
+export const sleep = (ms: number) => {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+};
diff --git a/webui/react/src/hooks/usePolling.ts b/webui/react/src/hooks/usePolling.ts
index cb2da09a..3299786d 100644
--- a/webui/react/src/hooks/usePolling.ts
+++ b/webui/react/src/hooks/usePolling.ts
@@ -2,7 +2,7 @@ import { useCallback, useEffect, useRef } from 'react';
 
 import { isAsyncFunction } from 'utils/data';
 
-const DEFAULT_DELAY = 5000;
+const DEFAULT_DELAY = 2000;
 
 type PollingFn = (() => Promise<void>) | (() => void);
 
diff --git a/webui/react/src/services/apiBuilder.ts b/webui/react/src/services/apiBuilder.ts
index 7f1392db..23d2074b 100644
--- a/webui/react/src/services/apiBuilder.ts
+++ b/webui/react/src/services/apiBuilder.ts
@@ -1,5 +1,6 @@
 import axios, { AxiosResponse, CancelToken, Method } from 'axios';
 
+import { sleep } from 'Dev';
 import handleError, { DaError, ErrorLevel, ErrorType, isDaError } from 'ErrorHandler';
 import { isAuthFailure } from 'services/api';
 import * as DetSwagger from 'services/api-ts-sdk';
@@ -53,11 +54,14 @@ export const processApiError = (name: string, e: Error): DaError => {
   });
 };
 
+let c = 0;
 export function generateApi<Input, Output>(api: Api<Input, Output>) {
   return async function(params: Input & { cancelToken?: CancelToken }): Promise<Output> {
+    const myC = c++;
     const httpOpts = api.httpOptions(params);
 
     try {
+      console.log('started req', myC, api.name, new Date());
       const response = await http.request({
         cancelToken: params.cancelToken,
         data: httpOpts.body,
@@ -66,6 +70,9 @@ export function generateApi<Input, Output>(api: Api<Input, Output>) {
         url: httpOpts.url as string,
       });
 
+      await sleep(10000);
+      console.log('finished req', myC, api.name, new Date());
+
       return api.postProcess ? api.postProcess(response) : response.data as Output;
     } catch (e) {
       processApiError(api.name, e);

```

**Review by hamidzr — COMMENTED** [review]
hmm I might be making a mistake but I think we're not getting the expected behavior :thinking:

**Comment by hkang1** [comment]
@hamidzr the `async` function works in the `usePolling` hook but the trick is to make sure that there is a `useCallback` wrapped around the async polling function. Otherwise the async polling function gets recreated each time during a re-render and will reset the polling mechanism with the new function. Here's is a sample diff where it is applied on the clusters page as an example.

```
diff --git a/webui/react/src/pages/Cluster.tsx b/webui/react/src/pages/Cluster.tsx
index 2f1db934..6086a7d8 100644
--- a/webui/react/src/pages/Cluster.tsx
+++ b/webui/react/src/pages/Cluster.tsx
@@ -1,4 +1,4 @@
-import React, { useMemo } from 'react';
+import React, { useCallback, useMemo } from 'react';
 
 import Grid from 'components/Grid';
 import Message from 'components/Message';
@@ -6,9 +6,16 @@ import Page from 'components/Page';
 import ResourceChart from 'components/ResourceChart';
 import Spinner from 'components/Spinner';
 import Agents from 'contexts/Agents';
+import usePolling from 'hooks/usePolling';
 import { Resource, ResourceType } from 'types';
 import { categorize } from 'utils/data';
 
+const doStuff = () => {
+  return new Promise(resolve => {
+    setTimeout(() => resolve('hello'), 10000);
+  });
+};
+
 const Cluster: React.FC = () => {
   const agents = Agents.useStateContext();
 
@@ -23,6 +30,13 @@ const Cluster: React.FC = () => {
 
   const availableResourceTypes = Object.keys(availableResources);
 
+  const fetchStuff = useCallback(async () => {
+    const response = await doStuff();
+    console.log('response', response);
+  }, []);
+
+  usePolling(fetchStuff, { delay: 1234 });
+
   if (!agents.data) {
     return <Spinner />;
   } else if (agents.data.length === 0) {
```

**Comment by hkang1** [comment]
@hamidzr I updated the `usePolling` hook to convert the polling function to `async` function if it's not one already. Any function can get wrapped into an async syntax, but only functions that return a promise will truly be blocked during an `await` call.

**Comment by hkang1** [comment]
This is being addressed in [PR 11316](https://github.com/determined-ai/determined/pull/1316)

**Comment by hkang1** [comment]
This is being addressed in [PR 11316](https://github.com/determined-ai/determined/pull/1316)

**Pull request closed without merging**